### DISCUSSION TITLE
Failing to add HelmRepository - failed to cache index - remote error: tls: internal error

### DISCUSSION THREAD

**Discussion by den-is** [fluxcd/flux2#3148] [Q&A] [answered]
I have fresh flux installation on a fresh on-prem k8s cluster.

I'm trying to add HelmRepository
```yaml
apiVersion: source.toolkit.fluxcd.io/v1beta2
kind: HelmRepository
metadata:
  name: ingress-nginx
  namespace: flux-system
spec:
  interval: 30m
  url: https://kubernetes.github.io/ingress-nginx
```
It is always failing with this error:
```
flux get sources helm
NAME            REVISION        SUSPENDED       READY   MESSAGE                                                         
ingress-nginx                   False           False   failed to fetch Helm repository index: failed to cache index to temporary file: Get "https://kubernetes.github.io/ingress-nginx/index.yaml": remote error: tls: internal error
```

```
k -n flux-system describe helmrepository/ingress-nginx
Name:         ingress-nginx
Namespace:    flux-system
Labels:       kustomize.toolkit.fluxcd.io/name=infrastructure
              kustomize.toolkit.fluxcd.io/namespace=flux-system
Annotations:  <none>
API Version:  source.toolkit.fluxcd.io/v1beta2
Kind:         HelmRepository
Metadata:
  Creation Timestamp:  2022-09-27T08:14:27Z
  Finalizers:
    finalizers.fluxcd.io
  Generation:  1
  Managed Fields:
    API Version:  source.toolkit.fluxcd.io/v1beta2
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:labels:
          f:kustomize.toolkit.fluxcd.io/name:
          f:kustomize.toolkit.fluxcd.io/namespace:
      f:spec:
        f:interval:
        f:url:
    Manager:      kustomize-controller
    Operation:    Apply
    Time:         2022-09-27T08:14:27Z
    API Version:  source.toolkit.fluxcd.io/v1beta2
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
          .:
          v:"finalizers.fluxcd.io":
    Manager:      source-controller
    Operation:    Update
    Time:         2022-09-27T08:14:27Z
    API Version:  source.toolkit.fluxcd.io/v1beta2
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        f:conditions:
    Manager:         source-controller
    Operation:       Update
    Subresource:     status
    Time:            2022-09-27T08:14:28Z
  Resource Version:  852125
  UID:               142444ef-ee2a-4107-8bcf-0d4e8188b29f
Spec:
  Interval:  30m
  Provider:  generic
  Timeout:   60s
  URL:       https://kubernetes.github.io/ingress-nginx
Status:
  Conditions:
    Last Transition Time:  2022-09-27T08:14:11Z
    Message:               no artifact for resource in storage
    Observed Generation:   1
    Reason:                NoArtifact
    Status:                True
    Type:                  Reconciling
    Last Transition Time:  2022-09-27T08:14:12Z
    Message:               failed to fetch Helm repository index: failed to cache index to temporary file: Get "https://kubernetes.github.io/ingress-nginx/index.yaml": remote error: tls: internal error
    Observed Generation:   1
    Reason:                Failed
    Status:                False
    Type:                  Ready
    Last Transition Time:  2022-09-27T08:14:12Z
    Message:               failed to fetch Helm repository index: failed to cache index to temporary file: Get "https://kubernetes.github.io/ingress-nginx/index.yaml": remote error: tls: internal error
    Observed Generation:   1
    Reason:                Failed
    Status:                True
    Type:                  FetchFailed
  Observed Generation:     -1
Events:
  Type     Reason  Age                    From               Message
  ----     ------  ----                   ----               -------
  Warning  Failed  5h53m (x8 over 5h56m)  source-controller  failed to fetch Helm repository index: failed to cache index to temporary file: Get "https://kubernetes.github.io/ingress-nginx/index.yaml": remote error: tls: internal error
```

```
flux check
► checking prerequisites
✔ Kubernetes 1.24.3 >=1.20.6-0
► checking controllers
✔ helm-controller: deployment ready
► ghcr.io/fluxcd/helm-controller:v0.24.0
✔ image-automation-controller: deployment ready
► ghcr.io/fluxcd/image-automation-controller:v0.25.0
✔ image-reflector-controller: deployment ready
► ghcr.io/fluxcd/image-reflector-controller:v0.21.0
✔ kustomize-controller: deployment ready
► ghcr.io/fluxcd/kustomize-controller:v0.28.0
✔ notification-controller: deployment ready
► ghcr.io/fluxcd/notification-controller:v0.26.0
✔ source-controller: deployment ready
► ghcr.io/fluxcd/source-controller:v0.29.0
► checking crds
✔ alerts.notification.toolkit.fluxcd.io/v1beta1
✔ buckets.source.toolkit.fluxcd.io/v1beta2
✔ gitrepositories.source.toolkit.fluxcd.io/v1beta2
✔ helmcharts.source.toolkit.fluxcd.io/v1beta2
✔ helmreleases.helm.toolkit.fluxcd.io/v2beta1
✔ helmrepositories.source.toolkit.fluxcd.io/v1beta2
✔ imagepolicies.image.toolkit.fluxcd.io/v1beta1
✔ imagerepositories.image.toolkit.fluxcd.io/v1beta1
✔ imageupdateautomations.image.toolkit.fluxcd.io/v1beta1
✔ kustomizations.kustomize.toolkit.fluxcd.io/v1beta2
✔ ocirepositories.source.toolkit.fluxcd.io/v1beta2
✔ providers.notification.toolkit.fluxcd.io/v1beta1
✔ receivers.notification.toolkit.fluxcd.io/v1beta1
✔ all checks passed
```

The source controller pod's log is showing the same error message.

Added `bitnami` HelmRepository source in the same directory next to ingress-nginx repo definition - and it worked flawlessly.
I have added other Private Github GitRepositories and doing HelmRelease from one of the repos just fine without any issues.

**Comment by stefanprodan** [comment]
We use that repo in e2e tests, here is my own cluster:

```yaml
apiVersion: source.toolkit.fluxcd.io/v1beta2
kind: HelmRepository
metadata:
  creationTimestamp: "2022-09-22T11:45:32Z"
  finalizers:
  - finalizers.fluxcd.io
  generation: 1
  labels:
    kustomize.toolkit.fluxcd.io/name: infra-controllers
    kustomize.toolkit.fluxcd.io/namespace: flux-system
  name: ingress-nginx
  namespace: ingress-nginx
  resourceVersion: "1125934"
  uid: 20d8efeb-0d3a-415c-90cb-95495e26d546
spec:
  interval: 24h
  provider: generic
  timeout: 60s
  url: https://kubernetes.github.io/ingress-nginx
status:
  artifact:
    checksum: d57a1001c520c37f8a8bf564820f858f51c03760bd2cd1a1d0f2574e02a5756b
    lastUpdateTime: "2022-09-27T15:29:01Z"
    path: helmrepository/ingress-nginx/ingress-nginx/index-d57a1001c520c37f8a8bf564820f858f51c03760bd2cd1a1d0f2574e02a5756b.yaml
    revision: d57a1001c520c37f8a8bf564820f858f51c03760bd2cd1a1d0f2574e02a5756b
    size: 110994
    url: http://source-controller.flux-system.svc.cluster.local./helmrepository/ingress-nginx/ingress-nginx/index-d57a1001c520c37f8a8bf564820f858f51c03760bd2cd1a1d0f2574e02a5756b.yaml
  conditions:
  - lastTransitionTime: "2022-09-22T11:45:33Z"
    message: stored artifact for revision 'd57a1001c520c37f8a8bf564820f858f51c03760bd2cd1a1d0f2574e02a5756b'
    observedGeneration: 1
    reason: Succeeded
    status: "True"
    type: Ready
  - lastTransitionTime: "2022-09-22T11:45:33Z"
    message: stored artifact for revision 'd57a1001c520c37f8a8bf564820f858f51c03760bd2cd1a1d0f2574e02a5756b'
    observedGeneration: 1
    reason: Succeeded
    status: "True"
    type: ArtifactInStorage
  observedGeneration: 1
  url: http://source-controller.flux-system.svc.cluster.local./helmrepository/ingress-nginx/ingress-nginx/index.yaml
```

Do you use some L7 egress proxy that does MITM?

**Reply by den-is** [reply]
This is a plain on-prem fresh cluster running in my home-lab, home-lan without any firewalls or proxies.
K8S cluster is running control and worker nodes in VirtualBox, they are RockyLinux9

Just curious how to troubleshoot that tls error, especially when i don't have sudo perms in the controller container

output that includes bitnami helmrepo that was installed way after initial ingress-nginx attempts:
```
flux get sources helm
NAME            REVISION                                                                SUSPENDED       READY   MESSAGE                                           
bitnami         1b387cfea68e7d8602ad77752f745ff8d90516239aee39effc3bdd235e5d0107        False           True    stored artifact for revision '1b387cfea68e7d8602ad77752f745ff8d90516239aee39effc3bdd235e5d0107'
ingressnginx                                                                            False           False   failed to fetch Helm repository index: failed to cache index to temporary file: Get "https://kubernetes.github.io/ingress-nginx/index.yaml": remote error: tls: internal error
```

```
tree sources
sources
├── bitnami.yaml
├── kustomization.yaml
└── nginx.yaml
```

I have tried to rename files in just 'nginx.yaml' and HelmRepository into ingressnginx to remove excess `-` but this did not help either.

**Reply by hiddeco** [reply]
Can you `curl` that same address from a worker node without any issues?

**Reply by den-is** [reply]
yea, both worker nodes return yaml
```
[root@worker2 ~]# curl -I https://kubernetes.github.io/ingress-nginx/index.yaml
HTTP/2 200
server: GitHub.com
content-type: text/yaml
permissions-policy: interest-cohort=()
x-origin-cache: HIT
last-modified: Fri, 16 Sep 2022 22:38:11 GMT
access-control-allow-origin: *
etag: "6324fad3-1b192"
expires: Fri, 23 Sep 2022 22:54:12 GMT
cache-control: max-age=600
x-proxy-cache: HIT
x-github-request-id: 082D:54AC:995B5F:9E59FE:632E36E4
accept-ranges: bytes
date: Tue, 27 Sep 2022 17:13:50 GMT
via: 1.1 varnish
age: 460
x-served-by: cache-lin2290024-LIN
x-cache: HIT
x-cache-hits: 2
x-timer: S1664298831.927858,VS0,VE1
vary: Accept-Encoding
x-fastly-request-id: 29f16b382eba965a0e4873979a6e0ca67222a039
content-length: 110994
```

or with redirect 
```
root@master1 ~# curl -IL https://kubernetes.github.io/ingress-nginx
HTTP/2 301
server: GitHub.com
content-type: text/html
permissions-policy: interest-cohort=()
location: https://kubernetes.github.io/ingress-nginx/
x-github-request-id: 8FEE:03EB:8F3871:93BD77:63332FAF
accept-ranges: bytes
date: Tue, 27 Sep 2022 17:15:33 GMT
via: 1.1 varnish
age: 6
x-served-by: cache-lin2290028-LIN
x-cache: HIT
x-cache-hits: 1
x-timer: S1664298934.773629,VS0,VE5
vary: Accept-Encoding
x-fastly-request-id: 9558ba94e0068965ceeaa786a5ff31a3ae76f713
content-length: 162

HTTP/2 200
server: GitHub.com
content-type: text/html; charset=utf-8
permissions-policy: interest-cohort=()
x-origin-cache: HIT
last-modified: Fri, 16 Sep 2022 22:38:11 GMT
access-control-allow-origin: *
etag: "6324fad3-9bdc"
expires: Tue, 27 Sep 2022 17:25:33 GMT
cache-control: max-age=600
x-proxy-cache: MISS
x-github-request-id: 0831:E88C:26F7C3:27D75C:63332FB5
accept-ranges: bytes
date: Tue, 27 Sep 2022 17:15:33 GMT
via: 1.1 varnish
age: 0
x-served-by: cache-lin2290028-LIN
x-cache: MISS
x-cache-hits: 0
x-timer: S1664298934.855566,VS0,VE118
vary: Accept-Encoding
x-fastly-request-id: 3807d144a1957021c8dd1e00239a2ac41525af10
content-length: 39900
```

**Reply by hiddeco** [reply]
I assume you already have tried simply kicking the source-controller pod to see if this magically resolves things? The source-controller pod comes with `wget` which could be used for further debugging, e.g. `wget -O - https://kubernetes.github.io/ingress-nginx/index.yaml`.

**Reply by den-is** [reply]
recreating source-controller pod did not help
Thanks for the info about wget, before I was trying to somehow install/update curl ca-certificates in that pod.
wget failed:
```
~ $ wget -O - https://kubernetes.github.io/ingress-nginx/index.yaml
Connecting to kubernetes.github.io (199.59.243.222:443)
140180865473352:error:14094438:SSL routines:ssl3_read_bytes:tlsv1 alert internal error:ssl/record/rec_layer_s3.c:1543:SSL alert number 80
ssl_client: SSL_connect
wget: error getting response: Connection reset by peer
```
And of course, it is an ugly busybox version of wget without original options, such as `--secure-protocol`

**Reply by den-is** [reply]
A complete reinstall of flux2 on the same cluster did not work either.
Later will try to apply the same setup on some other k8s cluster (minikube, k3d, ...)

**Reply by den-is** [reply]
TL;DR: flux2 setup of these HelmRepositories **on minikube instance has worked as intended**.

Now I will completely reset my "lab" cluster which is a pure **k8s v1.24.3**, setup without any 3rd party installers working since 2017.
Recently it was switched to RockyLinux 9.0+. Fresh deploys (on other OS versions including v9) of it, were working for many months without any issues for different scenarios.

There are two possible "hopes", that I will be happy to see. :)
- Or there is a bug in some flux component running on that platform
- Or of course, hopefully, there is some major configuration/health issues with some component in that cluster is broken. That same cluster was running ArgoCD for some time, without any configuration.

**Reply by den-is** [reply]
Failure again on a completely fresh and blank k8s cluster
This time cluster is running the latest version for almost every core component:
```
RockyLinux 9.0
containerd 1.6.8
k8s v1.24.6
calico 3.24.1
coredns 1.10.0
Linux 5.14.0-70.26.1.el9_0.x86_64 #1 SMP PREEMPT Tue Sep 20 17:53:31 UTC 2022
openssl-libs-3.0.1-41.el9_0.x86_64.rpm
openssl-3.0.1-41.el9_0.x86_64.rpm 
```
Upgrading host OS packages did not help.
selinux is disabled :)

source-controller error msg.. slightly formatted json :)
```json
{
  "level": "error",
  "ts": "2022-09-28T17:38:31.951Z",
  "msg": "failed to fetch Helm repository index: failed to cache index to temporary file: Get \"https://kubernetes.github.io/ingress-nginx/index.yaml\": remote error: tls: internal error",
  "name": "ingress-nginx",
  "namespace": "flux-system",
  "reconciler kind": "HelmRepository",
  "annotations": null,
  "error": "Failed",
  "stacktrace": 
    "github.com/fluxcd/pkg/runtime/events.(*Recorder).Eventf
      github.com/fluxcd/pkg/runtime@v0.18.0/events/recorder.go:113
    github.com/fluxcd/source-controller/internal/reconcile/summarize.RecordContextualError
      github.com/fluxcd/source-controller/internal/reconcile/summarize/processor.go:49
    github.com/fluxcd/source-controller/internal/reconcile/summarize.(*Helper).SummarizeAndPatch
      github.com/fluxcd/source-controller/internal/reconcile/summarize/summary.go:182
    github.com/fluxcd/source-controller/controllers.(*HelmRepositoryReconciler).Reconcile.func1
      github.com/fluxcd/source-controller/controllers/helmrepository_controller.go:193
    github.com/fluxcd/source-controller/controllers.(*HelmRepositoryReconciler).Reconcile
      github.com/fluxcd/source-controller/controllers/helmrepository_controller.go:222
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
      sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:121
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
      sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:320
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
      sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273
    sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
      sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234"
}
```
How to debug that issue?

**Reply by den-is** [reply]
Almost all repos from this list are failing
```
wget -O - https://charts.jetstack.io
wget -O - https://containous.github.io/traefik-helm-chart
wget -O - https://argoproj.github.io/argo-helm
+ wget -O - https://charts.bitnami.com/bitnami
wget -O - https://kubernetes.github.io/ingress-nginx
wget -O - https://prometheus-community.github.io/helm-charts
wget -O - https://grafana.github.io/helm-charts
wget -O - https://groundhog2k.github.io/helm-charts/
wget -O - https://aws.github.io/eks-charts
wget -O - https://spotinst.github.io/spotinst-kubernetes-helm-charts
wget -O - https://assets.logdna.com/charts
wget -O - https://kubernetes-sigs.github.io/metrics-server
wget -O - https://stefanprodan.github.io/podinfo
wget -O - https://charts.jenkins.io
wget -O - https://kubernetes-sigs.github.io/nfs-ganesha-server-and-external-provisioner/
wget -O - https://www.selenium.dev/docker-selenium
+ wget -O - https://charts.karpenter.sh/
wget -O - https://charts.fluxcd.io
wget -O - https://kubernetes.github.io/dashboard/
```
Except bitnami and karpenter repos

**Reply by stefanprodan** [reply]
All these repos that are failing are hosted on GitHub Pages and have the same root CA.

**Reply by den-is** [reply]
I've built my own source-controller image with ~blackjack and ...~ with root user enabled.. manually loaded it into the cluster.
disable PodSecurity and Context restrictions... will see how it helps .......

**Reply by den-is** [reply]
idk, no luck...
```
curl -ILv https://kubernetes.github.io/ingress-nginx
*   Trying 199.59.243.222:443...
* Connected to kubernetes.github.io (199.59.243.222) port 443 (#0)
* ALPN: offers h2
* ALPN: offers http/1.1
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: none
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS alert, internal error (592):
* error:14094438:SSL routines:ssl3_read_bytes:tlsv1 alert internal error
* Closing connection 0
curl: (35) error:14094438:SSL routines:ssl3_read_bytes:tlsv1 alert internal error
```
```
wget -O - https://kubernetes.github.io/ingress-nginx
--2022-09-28 18:40:14--  https://kubernetes.github.io/ingress-nginx
Resolving kubernetes.github.io (kubernetes.github.io)... 199.59.243.222
Connecting to kubernetes.github.io (kubernetes.github.io)|199.59.243.222|:443... connected.
OpenSSL: error:14094438:SSL routines:ssl3_read_bytes:tlsv1 alert internal error
Unable to establish SSL connection.
```
```
openssl s_client -connect kubernetes.github.io:443
CONNECTED(00000003)
139767045167944:error:14094438:SSL routines:ssl3_read_bytes:tlsv1 alert internal error:ssl/record/rec_layer_s3.c:1543:SSL alert number 80
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 7 bytes and written 322 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
```
```
/ # libressl s_client -connect kubernetes.github.io:443
CONNECTED(00000003)
140254534105928:error:14004438:SSL routines:CONNECT_CR_SRVR_HELLO:tlsv1 alert internal error:tls13_lib.c:129:SSL alert number 80
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 7 bytes and written 316 bytes
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : 0000
    Session-ID:
    Session-ID-ctx:
    Master-Key:
    Start Time: 1664391761
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
---
```

**Reply by den-is** [reply]
ok all of these repos from container resolve to that strange IP:
```
curl -ILv https://charts.jetstack.io
*   Trying 199.59.243.222:443...
curl -Iv https://charts.jenkins.io
*   Trying 199.59.243.222:443...
```
While on the host system that is configured with google dnss
```
curl -ILv https://charts.jetstack.io
*   Trying 172.66.47.109:443...
curl -Iv https://charts.jenkins.io
*   Trying 185.199.108.153:443...
```
but nslookup in container resolves correctly
```
nslookup kubernetes.github.io
Server:         10.33.0.10
Address:        10.33.0.10:53

Non-authoritative answer:
Name:   kubernetes.github.io
Address: 185.199.109.153
Name:   kubernetes.github.io
Address: 185.199.108.153
Name:   kubernetes.github.io
Address: 185.199.110.153
Name:   kubernetes.github.io
Address: 185.199.111.153

Non-authoritative answer:
Name:   kubernetes.github.io
Address: 2606:50c0:8000::153
Name:   kubernetes.github.io
Address: 2606:50c0:8001::153
Name:   kubernetes.github.io
Address: 2606:50c0:8002::153
Name:   kubernetes.github.io
Address: 2606:50c0:8003::153
```

**curl with the --resolve flag finally works**
```
curl -IL https://kubernetes.github.io/ingress-nginx --resolve kubernetes.github.io:443:185.199.109.153
HTTP/2 301
server: GitHub.com
content-type: text/html
permissions-policy: interest-cohort=()
location: https://kubernetes.github.io/ingress-nginx/
x-github-request-id: D97E:388C:155FF28:1603BB3:63349AC4
accept-ranges: bytes
date: Wed, 28 Sep 2022 19:20:32 GMT
via: 1.1 varnish
age: 956
x-served-by: cache-lin2290034-LIN
x-cache: HIT
x-cache-hits: 1
x-timer: S1664392833.594547,VS0,VE5
vary: Accept-Encoding
x-fastly-request-id: 790c1edce3708375c2b6f9a3311e260fee368392
content-length: 162

HTTP/2 200
server: GitHub.com
content-type: text/html; charset=utf-8
permissions-policy: interest-cohort=()
last-modified: Wed, 28 Sep 2022 14:30:27 GMT
access-control-allow-origin: *
etag: "63345a83-9bdc"
expires: Wed, 28 Sep 2022 19:14:37 GMT
cache-control: max-age=600
x-proxy-cache: MISS
x-github-request-id: 0818:8407:158AF44:162D56C:63349AC5
accept-ranges: bytes
date: Wed, 28 Sep 2022 19:20:32 GMT
via: 1.1 varnish
age: 0
x-served-by: cache-lin2290034-LIN
x-cache: HIT
x-cache-hits: 1
x-timer: S1664392833.674797,VS0,VE116
vary: Accept-Encoding
x-fastly-request-id: a8b087441e1de330b74554cacfe934caee1a5ee4
content-length: 39900
```

**Reply by hiddeco** [reply]
What happens if you spin up a Pod with the Alpine image we use as our base image? Does this present the same weird resolvance of the GitHub.io domain? If that's the case, I am afraid something is attempting to hijack your connection, especially based on the Google results of that IP.

Also given the current information you have shared, and the various checks we have done on our own systems (and the popularity of the Helm repositories you listed among our users). I do find it (highly) unlikely it has to do with us.

**Reply by hiddeco** [reply]
~Interestingly enough, this same IP shows up in https://github.com/OpenVPN/openvpn3-linux/issues/151. Which makes it seem (together with the Google reports about the IP) that it's trying to go through (some sort of) a VPN.~

Scratch that, it's because xx.com is resolving to the same IP. In this case, hijacking continues to be the most probable cause. Strengthened by things like this:

```console
~ $ nslookup giithub.com
Server:		9.9.9.9
Address:	9.9.9.9:53

Non-authoritative answer:
Name:	giithub.com
Address: 199.59.243.222

Non-authoritative answer

~ $ nslookup www.gtihub.com
Server:		9.9.9.9
Address:	9.9.9.9:53

Non-authoritative answer:
www.gtihub.com	canonical name = 86335.bodis.com
Name:	86335.bodis.com
Address: 199.59.243.222

Non-authoritative answer:
www.gtihub.com	canonical name = 86335.bodis.com

~ $ nslookup www.gitbub.com
Server:		9.9.9.9
Address:	9.9.9.9:53

Non-authoritative answer:
www.gitbub.com	canonical name = 86335.bodis.com
Name:	86335.bodis.com
Address: 199.59.243.222

Non-authoritative answer:
www.gitbub.com	canonical name = 86335.bodis.com
```

**Comment by den-is** [comment, marked as answer]
Creating dummy nginx-alpine pod in different namespaces - same story
**Created dummy nginx-debian pod and it brought something new:**
```
nslookup kubernetes.github.io
Server:         10.33.0.10
Address:        10.33.0.10#53

Non-authoritative answer:
Name:   kubernetes.github.io.dev
Address: 199.59.243.222
---
nslookup kubernetes.github.io 10.33.0.10
Server:         10.33.0.10
Address:        10.33.0.10#53

Name:   kubernetes.github.io.dev
Address: 199.59.243.222
```
That made it a little bit clear. WTH is that `.dev`
CoreDNS using standard, common configuration fallbacks to hosts `/etc/resolv.conf`
My kubernetes hosts resolv.conf
```
cat /etc/resolv.conf
# Generated by NetworkManager
search dev
nameserver 8.8.8.8
nameserver 8.8.4.4

Disallowing NetworkManager to manage /etc/resolv.conf and manually editing it has fixed the issue.
```
Hi - `search dev` on the first line
So it is about proper configuration of NetworkManager and /etc/resolv.conf
Then luckily discovered that Kubespray project has everything documented with real snippets:
https://github.com/kubernetes-sigs/kubespray/blob/master/docs/dns-stack.md
